### PR TITLE
parameterize alpn in connect/4 and listen/2

### PR DESCRIPTION
**Pull request by z8674558**
enable
`quicer:listen(Port, [{alpn, ["sample2", "sample"]}])` then
`quicer:connect("localhost", Port, [{alpn, "sample"}], 5000)`

### CODE REVIEW

File: c_src/quicer_nif.c [hunk 1]
```diff
@@ -413,7 +415,8 @@ ErlNifResourceType *ctx_stream_t = NULL;
 
 const QUIC_REGISTRATION_CONFIG RegConfig
     = { "quicer_nif", QUIC_EXECUTION_PROFILE_LOW_LATENCY };
-const QUIC_BUFFER Alpn = { sizeof("sample") - 1, (uint8_t *)"sample" };
+// @todo is 16 enough?
+const int MAX_ALPN = 16;
```
Comment (qzhuyan) [inline c_src/quicer_nif.c:419]: Should be uint16_t from https://tools.ietf.org/html/rfc7301

**Comment by qzhuyan** [comment]
Will be interested to see the behavior of msquic when alpn of client and server mismatch.

File: c_src/quicer_nif.c (hunk 1 above)
Comment (qzhuyan) [reply c_src/quicer_nif.c:419]: for now, it is ok to leave it 16.

File: c_src/quicer_config.c [hunk 2]
```diff
@@ -188,6 +207,49 @@ ClientLoadConfiguration(HQUIC *Configuration, bool Unsecure)
...
+  ERL_NIF_TERM head, tail;
+
+  if (!enif_get_list_cell(env, alpn_list, &head, &tail))
+    {
+      return false;
+    }
+
+  for(int i = 0; i < (int)(*alpn_buffer_length); i++) {
+
+  char str[PATH_MAX];
```
Comment (qzhuyan) [inline c_src/quicer_config.c:235]: Not sure PATH_MAX is correct, leave it as a @todo?

File: c_src/quicer_nif.h [hunk 3]
```diff
@@ -42,7 +42,7 @@ limitations under the License.
 extern HQUIC Registration;
 extern const QUIC_API_TABLE *MsQuic;
 extern const QUIC_REGISTRATION_CONFIG RegConfig;
-extern const QUIC_BUFFER Alpn;
+extern const int MAX_ALPN;
```
Comment (qzhuyan) [inline c_src/quicer_nif.h:45]: uint16_t

**Review by qzhuyan — APPROVED** [review]
LGTM just some small remarks.

**Comment by z8674558** [comment]
> Will be interested to see the behavior of msquic when alpn of client and server mismatch.

the client waits forever. I added the test for that.

File: test/quicer_SUITE.erl [hunk 4]
```diff
@@ -423,6 +426,40 @@ tc_peername_v4(Config) ->
...
+  end.
+
+tc_alpn_mismatch(Config) ->
+  Port = 4574,
+  Owner = self(),
+  Opts = lists:keyreplace(alpn, 1, default_listen_opts(Config), {alpn, ["no"]}),
+  {SPid, _Ref} = spawn_monitor(fun() -> conn_server_with(Owner, Port, Opts) end),
+  receive
+    listener_ready ->
+      spawn_monitor(fun() -> quicer:connect("localhost", Port, default_conn_opts(), 5000),
```
Comment (qzhuyan) [inline test/quicer_SUITE.erl:451]: We haven't done the code path for the closed connection (peer close).

I will fix it when the code is ready.

**Review by qzhuyan — APPROVED** [review]

**Pull request merged**